saleor/subscription/models.py

```diff
@@ -0,0 +1,126 @@
+from ..core.models import ModelWithMetadata
+from ..core.permissions import SubscriptionPermissions
+from ..core.utils.json_serializer import CustomJsonEncoder
+
+
+class SubscriptionQueryset(models.QuerySet):
+    def confirmed(self):
+        """Return non-draft subscriptions."""
+        return self.exclude(status=SubscriptionStatus.DRAFT)
+
+    def draft(self):
+        """Return draft subscriptions. """
+        return self.filter(status=SubscriptionStatus.DRAFT)
+
+    def active(self):
+        """ Return active subscriptions."""
+        return self.filter(status=SubscriptionStatus.ACTIVE)
+
+    def paused(self):
+        """ Return active subscriptions."""
+        return self.filter(status=SubscriptionStatus.PAUSED)
+
+    def ended(self):
+        """ Return active subscriptions."""
+        return self.filter(status=SubscriptionStatus.ENDED)
+
+
+class Subscription(ModelWithMetadata):
+    rrule = models.TextField()
+    start_date = models.DateTimeField()
+    upcoming_order_creation_date = models.DateTimeField(null=True)
+    upcoming_delivery_date = models.DateTimeField(null=True)
+    created = models.DateTimeField(default=now, editable=False)
+    status = models.CharField(
+        max_length=32, default=SubscriptionStatus.DRAFT, choices=SubscriptionStatus.CHOICES
+    )
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        blank=True,
+        null=True,
+        related_name="subscriptions",
+        on_delete=models.PROTECT,
+    )
+    billing_address = models.ForeignKey(
+        Address, related_name="+", editable=False, null=True, on_delete=models.SET_NULL
+    )
+    shipping_address = models.ForeignKey(
+        Address, related_name="+", editable=False, null=True, on_delete=models.SET_NULL
+    )
+    token = models.CharField(max_length=36, unique=True, blank=True)
+    customer_note = models.TextField(blank=True, default="")
+    ended_with_reason = models.TextField(blank=True, default="")
+    variant = models.ForeignKey(
+        "product.ProductVariant",
+        on_delete=models.PROTECT,
+        blank=True,
+        null=True,
+    )
+    quantity = models.IntegerField(validators=[MinValueValidator(1)])
+    objects = SubscriptionQueryset.as_manager()
+
+    class Meta:
+        ordering = ("-pk",)
+        permissions = ((SubscriptionPermissions.MANAGE_SUBSCRIPTIONS.codename, "Manage subscriptions."),)
+
+    def __repr__(self):
+        return "<Subscription #%r>" % (self.id,)
+
+    def __str__(self):
+        return "#%d" % (self.id,)
+
+    def save(self, *args, **kwargs):
+        if not self.token:
+            self.token = str(uuid4())
+        return super().save(*args, **kwargs)
+
+    def end(self, *args, **kwargs):
+        self.status=SubscriptionStatus.ENDED
+        if 'ended_with_reason' in kwargs:
+            self.ended_with_reason=kwargs["ended_with_reason"]
+
+    def get_customer_email(self):
+        return self.user.email if self.user else self.user_email
```

Comment:
`user_email` field does not exist in this model. Guest users won't be allowed to subscribe.

Reply:
Guest users should not be able to subscribe rt?

Reply:
removing the method